Add Panther algorithm per #3849

mrecachinas opened the pull request:
This addresses #3849.

A couple remaining questions:

* Is there a better home for `generate_random_paths`? It’s a generic Markovian random walk that could be used elsewhere. Alternatively, if this already exists or there's a more efficient approach, I'm all ears.
* I used `numpy` for a bit of this. Is there a desire to have a pure Python version as well (like `simrank_similarity` vs `simrank_similarity_numpy`)?
* Should `n_choose_k` be elsewhere or perhaps should we use `scipy.special.binom` instead?
* Do we want Panther++ added (requires a KD-tree, which we can use from `scipy.spatial`)?

pep8speaks commented:
Hello @mrecachinas! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`networkx/algorithms/similarity.py`](https://github.com/networkx/networkx/blob/817983257515493eba9f28b5c42fd00a5bd0a878/networkx/algorithms/similarity.py):

> [Line 1427:53](https://github.com/networkx/networkx/blob/817983257515493eba9f28b5c42fd00a5bd0a878/networkx/algorithms/similarity.py#L1427): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\s'
> [Line 1461:1](https://github.com/networkx/networkx/blob/817983257515493eba9f28b5c42fd00a5bd0a878/networkx/algorithms/similarity.py#L1461): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1489:33](https://github.com/networkx/networkx/blob/817983257515493eba9f28b5c42fd00a5bd0a878/networkx/algorithms/similarity.py#L1489): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1504:70](https://github.com/networkx/networkx/blob/817983257515493eba9f28b5c42fd00a5bd0a878/networkx/algorithms/similarity.py#L1504): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

mrecachinas commented:
There's an issue where node names must be the node's index... fixing this currently.

dschult commented:
```n_choose_k``` is nicely expressed in Python 3.8+ as ```math.comb(n, k)```.
But we probably need to keep something around until we stop supporting Python 3.7.

There is only a reason to keep a ```numpy``` version and plain python if there is an advantage to each approach. For example, the algorithm used might be different with one using matrix manipulations and one using graph search techniques, or similar.

Thanks for this!  I think there could be demand for Panther++. Let me think about where ```generate_random_paths``` might fit better.  

small tweaks -- typo "ictionary" in docs for ```generate_random_paths```.  And could the example do soemthing with random_path to show that it has two outputs and what someone might want to do with them?  Perhaps you could return just the paths and have an optional argument ```index_map``` that could be filled if supplied. ??

mrecachinas commented:
@dschult Fixed the typo and moved `index_map` to be an optional kwarg that, if provided, will be populated. I also added an example in the docstring illustrating that.

Let me know if there's anything else. I'll finish my implementation of Panther++.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1386,3 +1393,249 @@ def simrank_similarity_numpy(
+    path_length : int
+        How long the randomly generated paths should be (``T`` in [1]_)
+    c : float
+        A universal positive constant. Defaults to 0.5 per [1]_
```
c: float (default = 0.5)
    A universal positive constant used to scale the number of sample random paths to generate.

mrecachinas replied:
Addressed in 221da98.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1386,3 +1393,249 @@ def simrank_similarity_numpy(
+    ----------
+    G : NetworkX graph
+        A NetworkX graph
+    v : node
```
Call the node source instead of v?

mrecachinas replied:
Addressed in 81e2307.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1386,3 +1393,249 @@ def simrank_similarity_numpy(
+
+def panther_similarity(G, v, k=5, path_length=5, c=0.5, delta=0.1, eps=None):
+    """Returns the Panther similarity of nodes in the graph ``G`` to node ``v``.
+    Panther is a similarity metric that says "two objects are considered
```
add blank line after single-line function description (used for links to this function in Sphinx doc creation).

mrecachinas replied:
Addressed in 81e2307.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1386,3 +1393,249 @@ def simrank_similarity_numpy(
+
+
+def generate_random_paths(G, sample_size, path_length=5, index_map=None):
+    """Randomly generate ``sample_size`` paths of length ``path_length``.
```
Would ```generate_random_paths``` be better as a generator yielding paths?

mrecachinas replied:
That makes sense to me. Addressed in 49302de.

storopoli commented:
Any news guys? Will it be released in the next version?

dschult reviewed (commented):
I just have a few nits to pick. And, yes...  This will definitely get into the next release.

- Could we keep n_choose_k an unadvertised function by taking it out of ```__all__```? It really should be ```math.comb(n, k)``` but we have to wait for Python v3.8 to make that work. Perhaps a note in the doc_string to that effect is reasonable too.

- Add hooks to the docs in ```doc/reference/algorithms/similarity.rst``` by just adding the names of the new functions to that file.  That'll pull the doc_strings into the sphinx doc system automagically.

- I believe the ```setdefault``` method is advised against. Can you replace that idiom with something like:
```python
if node in index_map:
    index_map[node].add(path_index)
else:
    indexmap[node] = {path_index}
```

- You add ```1/ sample_size``` repeatedly does it make any difference in speed if you store ```1/sample_size``` once snd then use it vs recomputing in the loop?  I assume the difference is small, but my ancient fortran background is making me twitch. :}  You can leave it or change... 

That's all I've got.  Nice job -- very clean and clear code. :}

dschult commented:
Rats -- I didn't notice....  two things:  
1) you don't need the ```>>> import networkx as nx``` in the doc_string examples. That is run automagically by the testing software before the example code.
2) you can use ```n_choose_k``` without reference to ```nx.``` because it is in the current module.

So I think the errors go away by removing ```nx.``` from the ```nx.n_choose_k``` and for looks, please take out ```import networkx as nx```

Thanks!

mrecachinas commented:
@dschult Think I addressed everything. Let me know if there's anything else you would like changed!

Also, I'll submit another PR for Panther++ once I finish it.

storopoli commented:
@mrecachinas great job! Waiting for the panther++ algorithm also...

dschult commented:
Quick question:  In the docs, delta is:  
The probability that 𝑆 is not an epsilon-approximation to (R, phi)

What are R and phi?

You can answer here-- but if you prefer, you can add more to that doc string.

storopoli commented:
@dschult according to the original Panther article, $R$ is the number of random paths and $\phi$  is the probability that an element sampled from a set A $\subseteq$ D, where D is the domain.

ref: Zhang, J., Tang, J., Ma, C., Tong, H., Jing, Y., & Li, J. (2015). Panther: Fast Top-k Similarity Search on Large Networks. Proceedings of the 21th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, 1445–1454. https://doi.org/10.1145/2783258.2783267

dschult commented:
Good Thanks!   

But what is S... ?   I don't think these symbols are defined in the doc_string.... only in the article.
The code refers to "S", but it is an array of values, which is hard to see how to be epsilon close to a 2-tuple (R, phi)...

mrecachinas commented:
Sorry - S is just the similarity.

mrecachinas commented:
@jarrodmillman Let me know if I've addressed your change requests.

jarrodmillman commented:
@mrecachinas Sorry for the delay.  Ross or I will take a look very soon!  We are planning to release 2.6 soonish and I want to get this in first.  Thanks for your patience.

Despite our slow response time, we would also like to get Panther++ in soon.  So I hope you are still interested in contributing it!

rossbar reviewed (approved):
I went ahead an made some small touch-ups in rossbar/networkx@0973646c. I tried to push these to your branch (to spare you the spam via the github suggestion feature) but was unable to do so, likely because it seems like your fork originated from somewhere other than `networkx/networkx`. I wasn't able to PR against your fork either, so if you could apply those changes via cherry-pick that'd be great. If you're not comfortable doing so, you can either:
 1) Add rossbar as a collaborator to your fork and I'll open a PR against your fork or
 2) We just put this in and I can make a follow-on PR

Generally this LGTM - I'm not very familiar with the algorithm so my opinion shouldn't be weighted too heavily but the implementation seems good! I too was struck by the question of whether there is a better place for `generate_random_paths`, but I don't have any ideas and I don't think that should be a sticking point. Thanks @mrecachinas !

mrecachinas commented:
@jarrodmillman I'll dust off and finish the Panther++ I started a few months ago.

@rossbar Ahh, I think I forked someone else's fork for a previous PR (adding SimRank, IIRC), so yeah that's probably the issue. I'll cherry-pick if you're good with it.
